Library/Homebrew/utils/github/api.rb

```diff
@@ -181,7 +181,7 @@ def self.credentials_error_message(response_headers, needed_scopes)
     end
 
     def self.open_rest(
-      url, data: nil, data_binary_path: nil, request_method: nil, scopes: [].freeze, parse_json: true
+      url, data: nil, data_binary_path: nil, request_method: nil, scopes: [].freeze, parse_json: true, **options
```

Comment:
Please add named, keyword parameters instead of just an untyped Hash like `options`.

Reply:
I understand that explicit keyword parameters are necessary to be able to add meaningful type signatures but this is a somewhat complicated situation at the moment:

* `#curl_with_workarounds` has a gnarly method signature, with an `*args` splat (curl arguments like the URL, `--location`, etc.), eight keyword parameters, and an `**options` double splat (`#curl_args` arguments).
* `#curl_args` (which `#curl_with_workarounds` calls) has eight keyword parameters (different from `#curl_with_workarounds` eight keyword parameters).
* `#curl_output` (which `#open_rest` calls), simply uses an `**options` double splat to handle the keyword arguments for `#curl_with_workarounds` and `#curl_args`.
* Though I only mentioned `debug` above, what we really need to do in livecheck is to pass `**Strategy::DEFAULT_CURL_OPTIONS` to `#open_rest`. This is a hash with eight values, five of which are arguments for `#curl_with_workarounds` and three are arguments for `#curl_args`.
* The `GitHub` module doesn't have any type signatures at the moment and there are ten methods using a double splat there.

To avoid using `**options`, we would have to add 16 keyword parameters (i.e., the existing parameters for `#curl_with_workarounds` and `#curl_args`) to any method where we want to be able to change those parameters. We would also then have to explicitly pass those arguments in the `#curl_output` call. If we add only the keyword parameters we currently  need, this would still be eight and we would have to modify this method again if we ever add any more to `DEFAULT_CURL_OPTIONS`. This adds a lot of overhead and wouldn't immediately benefit us because `GitHub` (and `Utils::Curl`) methods don't have type signatures yet.

In the longer-term, one potential solution may be to create [separate] objects/structs for the keyword arguments for `#curl_with_workarounds` and `#curl_args`. With that setup, we could expect a specific type (e.g., `CurlArgsOptions`) and get rid of the related double splats. However, that's just a thought (I haven't tested the idea) and there may be challenges in making it work.

With that in mind, I agree that this isn't what we _should_ be doing but it feels like the only reasonable way of dealing with this until the situation is improved in `Utils::Curl`.

Reply:
> This seems better for a typing perspective. In this case specifically, though, presumably you don't need to pass all 16? Can you consider just adding the ones you need?

We currently need eight for livecheck (across `#curl_with_workarounds` and `#curl_args`). I can add just those but my concern is that it will bloat the `#open_rest` method signature and `#curl_output` call without buying us anything meaningful until `Utils::Curl` and `GitHub` have type signatures. I don't think anyone will attempt to type `GitHub` until after `Utils::Curl` is typed and that won't happen until the `Curl` arguments situation is improved. When that happens, we will need to replace either `**options` or the explicit keyword parameters in `#open_rest` and it's a bit easier to replace `**options`.

That said, it's not a huge issue either way, so if you really want to move forward with explicit keyword parameters instead of `**options` here, I'll update this accordingly.